File: scripts/checker.go
package main

import (
	"bufio"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"
)

// --- КОНФИГУРАЦИЯ ---
const (
	MaxNodes     = 50
	MinPing      = 0    // ФИКС: убрали нижний порог, быстрые узлы теперь не отсекаются
	MaxPing      = 3000 // ФИКС: увеличили верхний порог
	Threads      = 100
	OutputMobile = "configs/kudryash0vv_YKTFLOW_mobile.txt"
	OutputFull   = "configs/kudryash0vv_YKTFLOW_1.txt"
	SourceFile   = "data/sources_mobile.txt"
)

type Result struct {
	Raw     string
	Ping    int64
	Country string
	IP      string
	Flag    string
}

type GeoIP struct {
	CountryCode string `json:"countryCode"`
}

// --- УТИЛИТЫ ---

func getFlag(code string) string {
    if len(code) != 2 {
        return "🌍"
    }
    code = strings.ToUpper(code)
    // ФИКС: явно кастуем к rune перед сложением
    return string(rune(code[0])+127397) + string(rune(code[1])+127397)
}

func fetchCountry(ip string) (string, string) {
	client := &http.Client{Timeout: 2 * time.Second}
	// ФИКС: был неправильный URL без /json/ — GeoIP всегда возвращал пустой ответ
	resp, err := client.Get("http://ip-api.com/json/" + ip + "?fields=countryCode")
	if err != nil {
		return "UN", "🌍"
	}
	defer resp.Body.Close()

	var g GeoIP
	json.NewDecoder(resp.Body).Decode(&g)
	if g.CountryCode == "" {
		return "UN", "🌍"
	}
	return g.CountryCode, getFlag(g.CountryCode)
}

// --- ЯДРО ПРОВЕРКИ ---

func checkNode(raw string) (*Result, bool) {
	u, err := url.Parse(raw)
	if err != nil || u.Hostname() == "" {
		return nil, false
	}

	host := u.Hostname()
	port := u.Port()
	if port == "" {
		port = "443"
	}
	target := net.JoinHostPort(host, port)

	start := time.Now()

	// 1. TCP Dial
	dialer := &net.Dialer{Timeout: 3 * time.Second}
	conn, err := dialer.Dial("tcp", target)
	if err != nil {
		return nil, false
	}
	defer conn.Close()

	// 2. TLS/Reality Handshake
	security := u.Query().Get("security")
	if security == "tls" || security == "reality" {
		sni := u.Query().Get("sni")
		if sni == "" {
			sni = host
		}
		tlsConn := tls.Client(conn, &tls.Config{
			InsecureSkipVerify: true,
			ServerName:         sni,
		})
		tlsConn.SetDeadline(time.Now().Add(3 * time.Second))
		if err := tlsConn.Handshake(); err != nil {
			return nil, false
		}

		// ФИКС: теперь шлём реальный HTTP запрос и читаем ответ
		// Раньше был только Write без Read — сервер мог быть мёртвым
		httpReq := "GET / HTTP/1.1\r\nHost: " + sni + "\r\nConnection: close\r\n\r\n"
		tlsConn.SetWriteDeadline(time.Now().Add(1 * time.Second))
		if _, err := tlsConn.Write([]byte(httpReq)); err != nil {
			return nil, false
		}

		// Читаем хотя бы первые байты ответа — значит сервер живой
		tlsConn.SetReadDeadline(time.Now().Add(2 * time.Second))
		buf := make([]byte, 64)
		n, err := tlsConn.Read(buf)
		if err != nil || n == 0 {
			return nil, false
		}

	} else {
		// Для non-TLS узлов (например, ws без tls) — просто проверяем TCP
		conn.SetDeadline(time.Now().Add(2 * time.Second))
		buf := make([]byte, 1)
		conn.Read(buf) // Игнорируем ошибку — нам важен сам факт соединения
	}

	ping := time.Since(start).Milliseconds()

	// 3. Фильтрация по пингу
	if ping < MinPing || ping > MaxPing {
		return nil, false
	}

	countryCode, flag := fetchCountry(host)

	return &Result{
		Raw:     raw,
		Ping:    ping,
		Country: countryCode,
		IP:      host,
		Flag:    flag,
	}, true
}

// --- ОБРАБОТКА ДАННЫХ ---

func process() {
	fmt.Println("🧊 YKTFLOW Engine v9.6 | Инициализация...")

	file, err := os.Open(SourceFile)
	if err != nil {
		fmt.Println("❌ Ошибка: Создайте файл data/sources_mobile.txt")
		return
	}
	defer file.Close()

	// Сбор ссылок
	var sources []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if strings.HasPrefix(line, "http") {
			sources = append(sources, line)
		}
	}

	// Извлечение конфигов
	re := regexp.MustCompile(`(vless|vmess|trojan|ss)://[^\s"']+`)
	uniqueMap := make(map[string]struct{})
	httpClient := &http.Client{Timeout: 15 * time.Second}

	fmt.Printf("📂 Загрузка из %d источников...\n", len(sources))
	for _, urlLink := range sources {
		resp, err := httpClient.Get(urlLink)
		if err != nil {
			continue
		}
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()

		matches := re.FindAllString(string(body), -1)
		for _, m := range matches {
			uniqueMap[m] = struct{}{}
		}
	}

	// Многопоточный чекер
	fmt.Printf("🚀 Анализ %d уникальных конфигов...\n", len(uniqueMap))
	var wg sync.WaitGroup
	resChan := make(chan *Result, len(uniqueMap))
	sem := make(chan struct{}, Threads)

	var mu sync.Mutex
	usedIPs := make(map[string]bool)

	for raw := range uniqueMap {
		wg.Add(1)
		go func(r string) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()

			if node, ok := checkNode(r); ok {
				mu.Lock()
				if !usedIPs[node.IP] {
					usedIPs[node.IP] = true
					resChan <- node
				}
				mu.Unlock()
			}
		}(raw)
	}

	wg.Wait()
	close(resChan)

	// Сортировка и выборка ТОП-50
	var finalNodes []*Result
	for r := range resChan {
		finalNodes = append(finalNodes, r)
	}

	sort.Slice(finalNodes, func(i, j int) bool {
		return finalNodes[i].Ping < finalNodes[j].Ping
	})

	if len(finalNodes) > MaxNodes {
		finalNodes = finalNodes[:MaxNodes]
	}

	// Сохранение
	saveResults(OutputMobile, finalNodes)
	saveResults(OutputFull, finalNodes)

	fmt.Printf("✨ Успех! В базу YKTFLOW добавлено %d узлов.\n", len(finalNodes))
}

func saveResults(path string, nodes []*Result) {
	f, _ := os.Create(path)
	defer f.Close()

	fmt.Fprintln(f, "# profile-title: kudryash0vv.YKTFLOW")
	fmt.Fprintln(f, "# subscription-userinfo: upload=0; download=0; total=885837004800; expire=1798675200")
	fmt.Fprintf(f, "# update: %s (YKT)\n", time.Now().Format("2006-01-02 / 15:04"))
	fmt.Fprintln(f, "# support: https://github.com")
	fmt.Fprintln(f, "")

	for i, n := range nodes {
		u, _ := url.Parse(n.Raw)
		// Формируем красивое имя: [01] 🇯🇵 | 240ms
		u.Fragment = fmt.Sprintf("[%02d] %s | %dms", i+1, n.Flag, n.Ping)
		fmt.Fprintln(f, u.String())
	}
}

func main() {
	// Создаём директории если нет
	os.MkdirAll("configs", 0755)
	os.MkdirAll("data", 0755)

	process()
}

File: scripts/checker.py
import subprocess
import requests
import json
import os
import time
import tempfile
import re
import threading
from urllib.parse import urlparse, parse_qs
from concurrent.futures import ThreadPoolExecutor, as_completed

# --- КОНФИГУРАЦИЯ ---
INPUT_FILE   = "configs/kudryash0vv_YKTFLOW_1.txt"        # файл от main.go
OUTPUT_FILE  = "configs/kudryash0vv_YKTFLOW_checked.txt"  # финальный файл
MAX_CONFIGS  = 200       # сколько рабочих конфигов собрать
THREADS      = 10        # параллельных проверок (не больше 15)
TIMEOUT      = 10        # секунд на проверку одного конфига
TEST_URL     = "https://www.gstatic.com/generate_204"     # лёгкий URL (204 = OK)
BASE_PORT    = 20000     # начальный порт для socks5
SINGBOX_BIN  = "sing-box"  # путь к бинарнику

# --- ГЛОБАЛЬНЫЕ ---
port_lock    = threading.Lock()
used_ports   = set()
results_lock = threading.Lock()
working      = []

def get_free_port():
    with port_lock:
        port = BASE_PORT
        while port in used_ports:
            port += 1
        used_ports.add(port)
        return port

def release_port(port):
    with port_lock:
        used_ports.discard(port)

# --- ПАРСЕРЫ URI ---

def parse_vless(uri):
    try:
        u = urlparse(uri)
        params = parse_qs(u.query)
        return {
            "scheme": "vless",
            "uuid": u.username,
            "host": u.hostname,
            "port": int(u.port or 443),
            "flow": params.get("flow", [""])[0],
            "security": params.get("security", ["none"])[0],
            "sni": params.get("sni", [u.hostname])[0],
            "fp": params.get("fp", ["chrome"])[0],
            "pbk": params.get("pbk", [""])[0],
            "sid": params.get("sid", [""])[0],
            "net_type": params.get("type", ["tcp"])[0],
            "path": params.get("path", ["/"])[0],
            "host_header": params.get("host", [u.hostname])[0],
            "raw": uri
        }
    except Exception:
        return None

def parse_vmess(uri):
    try:
        import base64
        b64 = uri[8:]
        b64 += "=" * (-len(b64) % 4)
        data = json.loads(base64.b64decode(b64).decode())
        return {
            "scheme": "vmess",
            "uuid": data.get("id"),
            "host": data.get("add"),
            "port": int(data.get("port", 443)),
            "alter_id": int(data.get("aid", 0)),
            "security": data.get("tls", ""),
            "sni": data.get("sni", data.get("add", "")),
            "net_type": data.get("net", "tcp"),
            "path": data.get("path", "/"),
            "host_header": data.get("host", data.get("add", "")),
            "raw": uri
        }
    except Exception:
        return None

def parse_trojan(uri):
    try:
        u = urlparse(uri)
        params = parse_qs(u.query)
        return {
            "scheme": "trojan",
            "password": u.username,
            "host": u.hostname,
            "port": int(u.port or 443),
            "sni": params.get("sni", [u.hostname])[0],
            "net_type": params.get("type", ["tcp"])[0],
            "path": params.get("path", ["/"])[0],
            "raw": uri
        }
    except Exception:
        return None

def parse_ss(uri):
    try:
        import base64
        u = urlparse(uri)
        userinfo = u.username or ""
        try:
            decoded = base64.b64decode(userinfo + "==").decode()
            method, password = decoded.split(":", 1)
        except Exception:
            method = userinfo
            password = u.password or ""
        return {
            "scheme": "ss",
            "method": method,
            "password": password,
            "host": u.hostname,
            "port": int(u.port or 8388),
            "raw": uri
        }
    except Exception:
        return None

def parse_config(raw):
    raw = raw.strip().split("#")[0]  # убираем fragment
    if raw.startswith("vless://"):
        return parse_vless(raw)
    elif raw.startswith("vmess://"):
        return parse_vmess(raw)
    elif raw.startswith("trojan://"):
        return parse_trojan(raw)
    elif raw.startswith("ss://"):
        return parse_ss(raw)
    return None

# --- ГЕНЕРАТОР SING-BOX CONFIG ---

def make_singbox_config(parsed, socks_port):
    scheme = parsed["scheme"]

    def tls_block(parsed):
        security = parsed.get("security", "")
        if security == "reality":
            return {
                "enabled": True,
                "server_name": parsed.get("sni", ""),
                "reality": {
                    "enabled": True,
                    "public_key": parsed.get("pbk", ""),
                    "short_id": parsed.get("sid", "")
                },
                "utls": {
                    "enabled": True,
                    "fingerprint": parsed.get("fp", "chrome")
                }
            }
        elif security == "tls":
            return {
                "enabled": True,
                "server_name": parsed.get("sni", ""),
                "insecure": True,
                "utls": {
                    "enabled": True,
                    "fingerprint": parsed.get("fp", "chrome")
                }
            }
        return None

    def transport_block(parsed):
        net_type = parsed.get("net_type", "tcp")
        if net_type == "ws":
            return {
                "type": "ws",
                "path": parsed.get("path", "/"),
                "headers": {"Host": parsed.get("host_header", "")}
            }
        elif net_type == "grpc":
            return {
                "type": "grpc",
                "service_name": parsed.get("path", "")
            }
        return None

    if scheme == "vless":
        outbound = {
            "type": "vless",
            "tag": "proxy",
            "server": parsed["host"],
            "server_port": parsed["port"],
            "uuid": parsed["uuid"],
            "flow": parsed.get("flow", "")
        }
        tls = tls_block(parsed)
        if tls:
            outbound["tls"] = tls
        transport = transport_block(parsed)
        if transport:
            outbound["transport"] = transport

    elif scheme == "vmess":
        outbound = {
            "type": "vmess",
            "tag": "proxy",
            "server": parsed["host"],
            "server_port": parsed["port"],
            "uuid": parsed["uuid"],
            "alter_id": parsed.get("alter_id", 0),
            "security": "auto"
        }
        if parsed.get("security") == "tls":
            outbound["tls"] = {
                "enabled": True,
                "server_name": parsed.get("sni", ""),
                "insecure": True
            }
        transport = transport_block(parsed)
        if transport:
            outbound["transport"] = transport

    elif scheme == "trojan":
        outbound = {
            "type": "trojan",
            "tag": "proxy",
            "server": parsed["host"],
            "server_port": parsed["port"],
            "password": parsed["password"],
            "tls": {
                "enabled": True,
                "server_name": parsed.get("sni", parsed["host"]),
                "insecure": True
            }
        }
        transport = transport_block(parsed)
        if transport:
            outbound["transport"] = transport

    elif scheme == "ss":
        outbound = {
            "type": "shadowsocks",
            "tag": "proxy",
            "server": parsed["host"],
            "server_port": parsed["port"],
            "method": parsed["method"],
            "password": parsed["password"]
        }
    else:
        return None

    return {
        "log": {"level": "error"},
        "inbounds": [{
            "type": "socks",
            "tag": "socks-in",
            "listen": "127.0.0.1",
            "listen_port": socks_port,
            "sniff": False
        }],
        "outbounds": [
            outbound,
            {"type": "direct", "tag": "direct"}
        ]
    }

# --- ПРОВЕРКА КОНФИГА ---

def check_config(raw):
    parsed = parse_config(raw)
    if not parsed:
        return None

    port = get_free_port()
    config = make_singbox_config(parsed, port)
    if not config:
        release_port(port)
        return None

    tmp = tempfile.NamedTemporaryFile(
        mode="w", suffix=".json", delete=False
    )
    json.dump(config, tmp, ensure_ascii=False)
    tmp.close()

    proc = None
    try:
        proc = subprocess.Popen(
            [SINGBOX_BIN, "run", "-c", tmp.name],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL
        )
        time.sleep(1.5)  # ждём пока sing-box поднимется

        proxies = {
            "http":  f"socks5h://127.0.0.1:{port}",
            "https": f"socks5h://127.0.0.1:{port}"
        }
        resp = requests.get(TEST_URL, proxies=proxies, timeout=TIMEOUT)

        if resp.status_code in (200, 204):
            return raw  # ✅ РАБОЧИЙ
        return None

    except Exception:
        return None
    finally:
        if proc:
            proc.terminate()
            try:
                proc.wait(timeout=3)
            except Exception:
                proc.kill()
        try:
            os.unlink(tmp.name)
        except Exception:
            pass
        release_port(port)

# --- СОХРАНЕНИЕ ---

def save_results(path, nodes):
    os.makedirs(os.path.dirname(path) or ".", exist_ok=True)
    with open(path, "w", encoding="utf-8") as f:
        f.write("# profile-title: kudryash0vv.YKTFLOW\n")
        f.write("# subscription-userinfo: upload=0; download=0; total=885837004800; expire=1798675200\n")
        f.write(f"# update: {time.strftime('%Y-%m-%d / %H:%M')} (YKT)\n")
        f.write("# support: https://github.com\n")
        f.write("# checked: REAL sing-box validation ✅\n\n")
        for i, raw in enumerate(nodes, 1):
            raw_clean = raw.split("#")[0]
            host = urlparse(raw_clean).hostname or "unknown"
            f.write(f"{raw_clean}#[{i:03d}] ✅ {host}\n")
    print(f"💾 Сохранено: {path}")

# --- MAIN ---

def main():
    print("🧊 YKTFLOW Checker | sing-box валидатор")
    print(f"📂 Вход: {INPUT_FILE}")
    print(f"🎯 Цель: {MAX_CONFIGS} рабочих конфигов\n")

    # Проверяем sing-box
    try:
        r = subprocess.run([SINGBOX_BIN, "version"],
                           capture_output=True, timeout=5)
        ver = r.stdout.decode().splitlines()[0] if r.stdout else "ok"
        print(f"✅ sing-box: {ver}\n")
    except Exception:
        print("❌ sing-box не найден!")
        print("   Установи: https://github.com/SagerNet/sing-box/releases")
        return

    # Читаем конфиги
    re_cfg = re.compile(r'(vless|vmess|trojan|ss)://[^\s]+')
    configs = []
    if not os.path.exists(INPUT_FILE):
        print(f"❌ Файл не найден: {INPUT_FILE}")
        return

    with open(INPUT_FILE, "r", encoding="utf-8") as f:
        for line in f:
            m = re_cfg.search(line.strip())
            if m:
                configs.append(m.group(0))

    configs = list(set(configs))
    print(f"📋 Конфигов для проверки: {len(configs)}")
    print(f"🔄 Потоков: {THREADS}\n")

    checked = 0
    found = 0
    stop_event = threading.Event()

    with ThreadPoolExecutor(max_workers=THREADS) as executor:
        futures = {executor.submit(check_config, raw): raw for raw in configs}

        for future in as_completed(futures):
            if stop_event.is_set():
                future.cancel()
                continue

            checked += 1
            result = future.result()

            if result:
                found += 1
                with results_lock:
                    working.append(result)
                host = urlparse(result.split("#")[0]).hostname or "?"
                print(f"  ✅ [{found:03d}/{MAX_CONFIGS}] {host}")

            if checked % 20 == 0:
                print(f"  🔍 Проверено: {checked}/{len(configs)} | Рабочих: {found}")

            if found >= MAX_CONFIGS:
                print(f"\n🎯 Цель достигнута!")
                stop_event.set()

    final = working[:MAX_CONFIGS]
    save_results(OUTPUT_FILE, final)
    print(f"\n✨ Готово! Рабочих конфигов: {len(final)}/{len(configs)} проверено")

if __name__ == "__main__":
    main()
